services/motion/builtin/builtin.go

```diff
@@ -231,25 +235,149 @@ func (ms *builtIn) MoveOnGlobe(
+	// TODO: if we disentangle creation of the kinematic base from the plan call we make we don't need this and can do this once
+	var kinematicBase kinematicbase.KinematicBase = nil
+
+	var positionMu sync.Mutex
+	// TODO: this has got to become a geo Point eventually but testing gets hard
+	var currentPosition *geo.Point
+
+	var replanFlag atomic.Bool
+	replanFlag.Store(true)
+
+	var successFlag atomic.Bool
+	successFlag.Store(false)
+
+	errChan := make(chan error)
+
+	cancelCtx, cancelFn := context.WithCancel(context.Background())
+	ms.cancelFn = cancelFn
+
+	positionPollingPeriod := time.Duration(1000/motionCfg.PositionPollingFreqHz) * time.Millisecond
+	obstaclePollingPeriod := time.Duration(1000/motionCfg.ObstaclePollingFreqHz) * time.Millisecond
+
+	startPolling := func(ctx context.Context, period time.Duration, fn func(context.Context)) {
+		ms.backgroundWorkers.Add(1)
+		goutils.ManagedGo(func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					startTime := time.Now()
+					fn(ctx)
+					timeElapsed := time.Since(startTime)
+					sleepTime := int(math.Max(0, float64(period-timeElapsed)))
+					if !goutils.SelectContextOrWait(ctx, time.Duration(sleepTime)) {
+						fmt.Println("cancelled")
+						return
+					}
+				}
+			}
+		}, ms.backgroundWorkers.Done)
+	}
+
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		for !successFlag.Load() && ctx.Err() == nil {
+			if replanFlag.Load() {
+				fmt.Println("replanning")
+
+				planMu.Lock()
+				currentPlan = nil
+				planMu.Unlock()
+
+				ms.cancelFn()
+				cancelCtx, cancelFn = context.WithCancel(ctx)
+				ms.cancelFn = cancelFn
+
+				plan, kb, err := ms.planMoveOnGlobe(
+					ctx,
+					componentName,
+					destination,
+					movementSensor,
+					obstacles,
+					kinematicsOptions,
+					extra,
+				)
+				if err != nil {
+					errChan <- err
+					return
+				}
+
+				planMu.Lock()
+				currentPlan = plan
+				kinematicBase = kb
+				planMu.Unlock()
+
+				replanFlag.Store(false)
+
+				startPolling(cancelCtx, positionPollingPeriod, func(ctx context.Context) {
+					// TODO: the function that actually monitors position
+					fmt.Println("position poll")
+					position, _, err := movementSensor.Position(cancelCtx, nil)
+					if err != nil {
+						errChan <- err
+						return
+					}
+					positionMu.Lock()
+					currentPosition = position
+					positionMu.Unlock()
+				})
+				startPolling(cancelCtx, obstaclePollingPeriod, func(ctx context.Context) {
+					fmt.Println("obstacle poll")
+					time.Sleep(time.Second)
+					replanFlag.Store(true)
+				})
+			}
+		}
+	}, ms.backgroundWorkers.Done)
 
 	// execute the plan
-	for i := 1; i < len(plan); i++ {
-		ms.logger.Info(plan[i])
-		if err := kb.GoToInputs(ctx, plan[i]); err != nil {
-			return false, err
+	var waypoint []referenceframe.Input
+	for !successFlag.Load() {
+		planMu.Lock()
+		if len(currentPlan) > 1 {
+			waypoint = []referenceframe.Input{
```

Comment:
This breaks for PTGs. Presumably we ought be able to callGoToInputs with the `currentPlan` item directly.

Reply:
Also, [opt] this can just be `{currentPlan[1][0].Value - currentPlan[0][0].Value}` no need to manually specify `Value:`

Reply:
Yeah this has got to change I just have it in now as a POC